frontend/src/app/modules/matches/matches-computations.js

```diff
@@ -68,3 +77,104 @@ export const computeWinRatio = (playerId, playerMatches) => {
   const wonMatchesCount = playerMatches.filter(match => didPlayerWin(playerId, match)).length
   return (playerMatches.length > 0) ? (wonMatchesCount / playerMatches.length) : 0
 }
+
+export const computeDays = matchChanges => {
+  let worstDayKey = ''
+  let bestDayKey = ''
+  const daysMap = {}
+  for (const match of matchChanges) {
+    const key = match.date.toLocaleDateString()
+    const value = Number(match.ratingChangeString)
+    if (!Object.hasOwnProperty.call(daysMap, key)) {
+      daysMap[key] = {
+        date: key,
+        value,
+      }
+    } else {
+      daysMap[key]['value'] += value
+    }
+
+    bestDayKey = updateMaxFromMap(daysMap[bestDayKey] ?? {}, 'value', daysMap[key]).date
+    worstDayKey = updateMinFromMap(daysMap[worstDayKey] ?? {}, 'value', daysMap[key]).date
+  }
+
+  return {
+    bestDay: daysMap[bestDayKey],
+    worstDay: daysMap[worstDayKey],
+  }
+}
+
+export const computeTeammates = (playerId, playerMatches) => {
+  const teammates = findSignificanPlayers(playerId, playerMatches, getTeamMates)
+  return {
+    mostFrequentTeammate: teammates.mostFrequentPlayer,
+    leastFrequentTeammate: teammates.leastFrequentPlayer,
+    mostSuccessTeammate: teammates.mostSuccessPlayer,
+    leastSuccessTeammate: teammates.leastSuccessPlayer,
+  }
+}
+
+export const computeOpponents = (playerId, playerMatches) => {
+  const opponents = findSignificanPlayers(playerId, playerMatches, getOpponents)
+  return {
+    mostFrequentOpponent: opponents.mostFrequentPlayer,
+    leastFrequentOpponent: opponents.leastFrequentPlayer,
+    mostSuccessOpponent: opponents.mostSuccessPlayer,
+    leastSuccessOpponent: opponents.leastSuccessPlayer,
+  }
+}
+
+const findSignificanPlayers = (playerId, playerMatches, playersProvider) => {
+  let mostFrequentPlayer = {}
+  let leastFrequentPlayer = {}
+  let mostSuccessPlayer = {}
+  let leastSuccessPlayer = {}
+  const playersMap = {}
+  for (const match of playerMatches) {
```

Comment:
as for computeDays - I would separate these in several passes for readability and scaling - one day we may need to have just a frequent player and we would need either to write it from scratch or somehow extract it from here with options to add bugs :(

Reply:
also same as with computeDays - maybe precompute data and then just run min, max algorithm - https://stackoverflow.com/questions/4020796/finding-the-max-value-of-an-attribute-in-an-array-of-objects - on a prepared array if possible

Reply:
preprocessing could also let you to avoid using playersProvider callback - you could have something like an array of all opponents and array of all teammates and then do computations on them. Again I think it would be more extensible in future.

Reply:
Did the thing with the min/max approach [here](https://github.com/salsita/foosball-rating/pull/114/commits/3d398530cdf75f980e6344e885e4f18dd919311a#diff-304dca28d92d292252e86b7e83fc1569) and [here](https://github.com/salsita/foosball-rating/pull/114/commits/3d398530cdf75f980e6344e885e4f18dd919311a#diff-08117b849ecc00af7c3c0d609dd86d31).

I don't like the idea of running through the matches and seperating players into structures of teammates and opponents and then running through them for actual computing. It would IMHO be less readable than a simple function that extracts either opponents or teammates but I am open to discussion in this...